File: cpp/200.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
    int numIslands(vector<vector<char>>& grid)
    {
        if (grid.size() == 0)
            return 0;
        int sum = 0;
        int m = grid.size() + 2;
        int n = grid[0].size() + 2;
        char** g = new char* [m];
        for (int i = 0; i < m; i++)
            g[i] = new char[n];
        for (int i = 0; i < m; i++)
        {
            for (int j = 0; j < n; j++)
            {
                if (i == 0 || j == 0 || i == m - 1 || j == n - 1)
                    g[i][j] = '0';
                else
                    g[i][j] = grid[i-1][j-1];
            }
        }
        for (int i = 1; i < m-1; i++)
        {
            for (int j = 1; j < n-1; j++)
            {
                if (g[i][j] == '1')
                {
                    digui(g, i, j);
                    sum++;
                }
            }
        }
        return sum;
    }

    void digui(char** grid,int i,int j)
    {
        if (grid[i][j] == '0'||grid[i][j]=='x')
            return;
        else if(grid[i][j]=='1')
            grid[i][j] = 'x';
        digui(grid, i - 1, j);
        digui(grid, i + 1, j);
        digui(grid, i, j - 1);
        digui(grid, i, j + 1);
    }
};

int main()
{
    Solution so;
    vector<vector<char>> grid;
    grid.resize(4);
    char c;
    for (int i = 0; i < 4; i++)
        for (int j = 0; j < 5; j++)
        {
            cin >> c;
            grid[i].push_back(c);
        }
    cout << so.numIslands(grid) << endl;

    return 0;
}

File: java/200.java
class Node{
    int x, y;

    public Node(int _x, int _y){
        x = _x;
        y = _y;
    }
}

class Solution {
    boolean[][] flag;
    int col, row;
    int[] X = {0, 0, 1, -1};
    int[] Y = {1, -1, 0, 0};


    boolean judge(int x, int y, char[][] grid){
        if(x<0 || x>=col || y<0 || y>=row){
            return false;
        }
        return grid[x][y] != '0' && !flag[x][y];
    }

    void BFS(int x, int y, char[][] grid){
        Node node = new Node(x, y);
        Queue<Node> queue = new LinkedList<>();
        queue.add(node);
        flag[x][y] = true;

        while (!queue.isEmpty()){
            Node tmp = queue.peek();
            queue.poll();

            for(int i=0; i<4; i++){
                int newX = tmp.x + X[i];
                int newY = tmp.y + Y[i];

                if(judge(newX, newY, grid)){
                    Node n = new Node(newX, newY);
                    queue.add(n);
                    flag[newX][newY] = true;
                }
            }
        }
    }

    public int numIslands(char[][] grid) {
        int cnt=0;
        col = grid.length;
        if(col == 0)
            return 0;
        row = grid[0].length;
        flag = new boolean[col][row];
        for(int i=0; i<col; i++){
            for(int k=0; k<row; k++)
                flag[i][k] = false;
        }

        for(int i=0; i<col; i++){
            for(int k=0; k<row; k++){
                if(grid[i][k] == '1' && flag[i][k] == false){
                    cnt++;
                    BFS(i, k, grid);
                }
            }
        }


        return cnt;
    }
}